pkitesting/src/main/java/io/netty/pkitesting/RevocationServer.java

```diff
@@ -0,0 +1,188 @@
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.pkitesting;
+
+import com.sun.net.httpserver.HttpServer;
+
+import java.io.OutputStream;
+import java.math.BigInteger;
+import java.net.InetAddress;
+import java.net.InetSocketAddress;
+import java.net.URI;
+import java.security.cert.X509Certificate;
+import java.time.Instant;
+import java.util.Collections;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.AtomicInteger;
+
+/**
+ * A simple HTTP server that serves Certificate Revocation Lists.
+ * <p>
+ * Issuer certificates can be registered with the server, and revocations of their certificates and be published
+ * and added to the revocation lists.
+ * <p>
+ * The server is only intended for testing usage, and runs entirely in a single thread.
+ *
+ * @implNote The CRLs will have the same very short life times, to minimize caching effects in tests.
+ * This currently means the time in the "this update" and "next update" fields are set to the same value.
+ */
+public final class RevocationServer {
+    private static RevocationServer instance;
+
+    private final HttpServer crlServer;
+    private final String crlBaseAddress;
+    private final AtomicInteger issuerCounter;
+    private final ConcurrentHashMap<X509Certificate, CrlInfo> issuers;
+    private final ConcurrentHashMap<String, CrlInfo> paths;
+
+    /**
+     * Get the shared revocation server instance.
+     * This will start the server, if it isn't already running, and bind it to a random port on the loopback address.
+     * @return The revocation server instance.
+     * @throws Exception If the server failed to start.
+     */
+    public static synchronized RevocationServer getInstance() throws Exception {
+        if (instance != null) {
+            return instance;
+        }
+        RevocationServer server = new RevocationServer();
+        server.start();
+        instance = server;
+        return server;
+    }
+
+    private RevocationServer() throws Exception {
+        // Use the JDK built-in HttpServer to avoid any circular dependencies with Netty itself.
+        crlServer = HttpServer.create(new InetSocketAddress(InetAddress.getLoopbackAddress(), 0), 0);
+        crlBaseAddress = "http://localhost:" + crlServer.getAddress().getPort();
+        issuerCounter = new AtomicInteger();
+        issuers = new ConcurrentHashMap<>();
+        paths = new ConcurrentHashMap<>();
```

Comment:
I think we need to ensure these maps can't grow forever as this instance will exists forever

Reply:
"Forever" still shouldn't be too long in this case, since as the class javadoc say:

> The server is only intended for testing usage